More jacksonising + changing representation to agreed upon docs

fppt opened the pull request:
@aelred @pluraliseseverythings we have an interesting side effect of our object wrappers. I will give you an example.

I changed the `Keyspace` json representation to be:

```
{
      "@id": "/kb/genealogy",
      "name": "genealogy",
      "types": "/kb/genealogy/type",
      "roles": "/kb/genealogy/role",
      "rules": "/kb/genealogy/rule",
      "graql": "/kb/genealogy/graql"
}
```

and that works very well. NOW, get getting a list of all the Keyspaces this is the representation we agreed on: 

```
{
  "@id": "/kb",
  "keyspaces": [
    {
      "@id": "/kb/genealogy",
      "name": "genealogy",
      "types": "/kb/genealogy/type",
      "roles": "/kb/genealogy/role",
      "rules": "/kb/genealogy/rule",
      "graql": "/kb/genealogy/graql"
    }
  ],
  "keyspace": "/kb/{keyspace}"
}
```

To achieve this I created the `Keyspaces` object which wraps the list plus the additional metadata. 

Now we get to an interesting problem. Serialising a json representation into one of our wrappers either `Keyspace` or `Keyspaces` throws the following exception (for example):

```
Exception in thread "main" com.fasterxml.jackson.databind.exc.UnrecognizedPropertyException: Unrecognized field "types" , not marked as ignorable
```

This is because the serialisation does not use the entire Json representation. For example when Serialising:

```
{
  "@id": "/kb",
  "keyspaces": [
    {
      "@id": "/kb/genealogy",
      "name": "genealogy",
      "types": "/kb/genealogy/type",
      "roles": "/kb/genealogy/role",
      "rules": "/kb/genealogy/rule",
      "graql": "/kb/genealogy/graql"
    }
  ],
  "keyspace": "/kb/{keyspace}"
}
```

The only thing we need is the name `"name": "genealogy",` the rest is all wrapped up metadata so we get the above exception. I workaround this by saying that partial usage of the Json representation is okay with `@JsonIgnoreProperties(ignoreUnknown = true)` I don't know if there is a neater way to do this BUT if there is please to tell me.

For the moment I will be continuing down this route.

aelred commented:
@fppt the "ignore properties" seems totally acceptable to me. If we think it's too permissive, we can selectively list them, like `@JsonIgnoreProperties({"@id", "types", "roles", "rules", "graql"})`.

fppt commented:
We likely to see them everywhere because most of our wrappers will serialise with only a part of the json representation. Just something to be aware of as I proceed

fppt commented:
**Note:** At this stage in the PR. List of keyspaces has changed from:

```
{
    keyspaces: ["a", "b", "c"]
}
```

To 

```
{
  "@id": "/kb",
  "keyspaces": [
    {
      "@id": "/kb/genealogy",
      "name": "genealogy",
      "types": "/kb/genealogy/type",
      "roles": "/kb/genealogy/role",
      "rules": "/kb/genealogy/rule",
      "graql": "/kb/genealogy/graql"
    }
  ],
  "keyspace": "/kb/{keyspace}"
}
```

fppt commented:
@marco-scoppetta Are there any dashboard breaking changes in this PR? 

I have changed the response of ConceptController which i believe you are not using but I also changed the response of the keyspace list. As stated above. I think that will break dashboard?

aelred commented on grakn-core/src/main/java/ai/grakn/util/Schema.java:
```diff
@@ -133,7 +134,8 @@ public static boolean isMetaLabel(Label label) {
         //Internal
-        SHARD(Vertex.class);
+        SHARD(Vertex.class),
+        CONCEPT(Concept.class);
```
New base type? :O

fppt replied:
It's used in the link class so I don't have to hard code the string "concept" there. The other way would be to have more Paths on `REST.WebPath` which would be redundant. This feels like the least evil appraoch.

aelred replied:
That's fair! I'd definitely have a comment to say that no vertex will _really_ have this type.

aelred commented on grakn-engine/src/main/java/ai/grakn/engine/controller/api/ConceptController.java:
```diff
@@ -0,0 +1,104 @@
+        spark.get(WebPath.CONCEPT_ID,  this::getConceptById);
+        spark.get(WebPath.TYPE_LABEL,  this::getSchemaByLabel);
+        spark.get(WebPath.RULE_LABEL,  this::getSchemaByLabel);
+        spark.get(WebPath.ROLE_LABEL,  this::getSchemaByLabel);
```
These are all the same right now. I guess we need to make the wrong one give some kind of 400 response.

aelred commented on grakn-core/src/main/java/ai/grakn/engine/Jacksonisable.java:
```diff
@@ -0,0 +1,36 @@
+ * @author Filipe Peliz Pinto Teixeira
+ */
+@JsonIgnoreProperties(value={ "@id", "roles", "types", "graql", "rules" }, allowGetters=true)
+public interface Jacksonisable {
```
Can we move `@id` into this interface?

fppt replied:
This will lead to a fun generic but I will give it a try.

fppt replied:
Actually scratch that. I am not sure how to do it because they all use it in different contexts. For example `ai.grakn.Keyspace` uses it but does not have a `@id` json property. It IS the `@id` property.

On the other hand, `ai.grakn.engine.controller.response.Concept` and `ai.grakn.engine.controller.response.Keyspace` use it and contain `@id` as a json property.

All 3 require the annotation provided by `Jacksonisable`. In other words if you want `@id` to be common place then another intr=erface might be needed and I am not sure we will gain much from that.

aelred commented on grakn-engine/src/main/java/ai/grakn/engine/controller/response/Role.java:
```diff
@@ -0,0 +1,57 @@
+    @JsonCreator
+    public static Role create(
+            @JsonProperty("id") ConceptId id,
+            @JsonProperty("@id") Link selfLink,
```
Maybe properties like `@id` can be derived instead of provided in the constructors.

fppt replied:
Can you elaborate on this? What do you mean?

aelred replied:
The `@id` `selfLink()` can (usually?) be derived from other properties, so you don't necessarily have to provide it in the constructor - just implement the method.

aelred commented on grakn-engine/src/main/java/ai/grakn/engine/controller/response/Thing.java:
```diff
@@ -0,0 +1,42 @@
+    public abstract Set<Link> keys();
+
+    @JsonProperty
+    public abstract Set<Link> relationships();
```
These links need to include the role (like `RolePlayer`)

aelred commented on grakn-engine/src/main/java/ai/grakn/engine/controller/response/AttributeType.java:
```diff
@@ -0,0 +1,53 @@
+ * @author Filipe Peliz Pinto Teixeira
+ */
+@AutoValue
+public abstract class AttributeType extends Type{
```
This needs a datatype, plus some other stuff (like regex).

fppt replied:
Yep you are right.

aelred commented on grakn-engine/src/main/java/ai/grakn/engine/controller/response/Link.java:
```diff
@@ -0,0 +1,74 @@
+                thing.keyspace().getValue(),
+                Schema.BaseType.CONCEPT.name().toLowerCase(Locale.getDefault()),
+                thing.getId().getValue());
+        return new AutoValue_Link(id);
```
Boring thing, but these should delegate to `create(String)` above instead of `AutoValue_Link`

aelred commented on grakn-engine/src/main/java/ai/grakn/engine/controller/response/Link.java:
```diff
@@ -0,0 +1,74 @@
+        if(schemaConcept.isRule()){
+            type = Schema.BaseType.RULE.name().toLowerCase(Locale.getDefault());
+        } else if(schemaConcept.isRole()){
+            type = Schema.BaseType.ROLE.name().toLowerCase(Locale.getDefault());
```
I guess we can simplify this to `if (concept.isType()) { TYPE } else { concept.baseType() }`

aelred commented on grakn-engine/src/main/java/ai/grakn/engine/controller/response/Rule.java:
```diff
@@ -0,0 +1,57 @@
+public abstract class Rule extends SchemaConcept{
+
+    @JsonProperty
+    public abstract String when();
```
These _could_ be `Pattern`, but there may be issues with that...

fppt replied:
That would be leaking Jackson annotations into more core objects and I am not a fan of that personally. `Keyspace` is okay because it's pretty much a wrapper by itself but `Pattern` contains a bit too much low level logic to allow the leak in my opinion.

aelred commented on grakn-engine/src/test/java/ai/grakn/engine/controller/SystemControllerTest.java:
```diff
@@ -98,17 +104,19 @@ public void whenCallingPutKBEndpoint_Return200_AndConfigInBody() throws JsonProc
-        when().get("/kb").then().body("", hasItem("myks"));
+        String content = when().get("/kb").thenReturn().body().asString();
+        Keyspaces keyspaces = objectMapper.readValue(content, Keyspaces.class);
+        assertThat(keyspaces.keyspaces(), not(empty()));
```
Apparently Rest Assured should handle this for us: https://github.com/rest-assured/rest-assured/wiki/Usage#object-mapping

```
Keyspaces keyspaces = get("/kb").as(Keyspaces.class);
```

aelred commented on grakn-engine/src/test/java/ai/grakn/engine/controller/api/ConceptControllerTest.java:
```diff
@@ -0,0 +1,162 @@
+        assertEquals(SC_OK, response.statusCode());
+        String content = response.thenReturn().body().asString();
+
+        assertEquals(wrapper, objectMapper.readValue(content, clazz));
```
Same as before: `assertEquals(wrapper, RestAssured.get(request).as(clazz));`

aelred commented on grakn-kb/src/main/java/ai/grakn/kb/internal/concept/RelationshipImpl.java:
```diff
@@ -118,6 +122,21 @@ public Relationship attributeRelationship(Attribute attribute) {
+
+        AttributeType[] keyTypesArray = keyTypes.toArray(new AttributeType[keyTypes.size()]);
+        return readFromReified((relationReified) -> relationReified.attributes(keyTypesArray));
+    }
```
This is duplicated from `ThingImpl`. Can you instead delegate to the reified instance? If the relationship is not reified, then it has no keys... I think.

aelred commented on grakn-engine/src/main/java/ai/grakn/engine/controller/response/ConceptBuilder.java:
```diff
@@ -0,0 +1,168 @@
+     * @return the response object wrapper {@link Concept}
+     */
+    public static <X extends Concept> Optional<X> build(ai.grakn.concept.Concept concept){
+        if(concept == null) return Optional.empty();
```
This looks like the only case where it returns `Optional.empty()`. I think it's clearer to handle it at the call-site:

```
repr = Optional.ofNullable(concept).map(ConceptBuilder::build);
```

But if you want to keep the null check, remember to mark the argument as `@Nullable` 👍

aelred commented on grakn-engine/src/main/java/ai/grakn/engine/controller/response/ConceptBuilder.java:
```diff
@@ -0,0 +1,168 @@
+        return Optional.of((X) response);
+    }
+
+    //TODO: This will scale poorly with super nodes. Need to introduce some sort of paging maybe?
```
Yeah... we might need a paged endpoint for relationships and roleplayers :(

fppt commented:
@aelred re-review please.

fppt commented:
@aelred this was merged (not complaining because I can continue working) without @marco-scoppetta confirming if these changes broke dashboard. So central may have a broken dashboard now.
